internal/state/relationship/capturer.go

```diff
@@ -0,0 +1,131 @@
+package relationship
+
+import (
+	v1 "k8s.io/api/core/v1"
+	discoveryV1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/pkg/sdk"
+)
+
+// Capturer captures relationships between Kubernetes objects and can be queried for whether a relationship exists for a given object.
+//
+// Currently, it only captures relationships between HTTPRoutes and Services and Services and EndpointSlices, but it can be extended to capture additional relationships.
+// The relationships between HTTPRoutes -> Services are many to 1, so these relationships are tracked using a counter. A Service relationship exists if at least one HTTPRoute references it.
+// An EndpointSlice relationship exists, if its Service owner is referenced by at least one HTTPRoute.
+type Capturer struct {
+	routesToServices    map[types.NamespacedName]map[types.NamespacedName]struct{}
```

Comment:
Type declarations would add to readability.

Reply:
Added